### DISCUSSION TITLE
How to programmatically mark a field as changed?

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#15046] [Archive]
Doing e2e tests with Cypress, I modify the content of a WYSIWYG field by program
```
        cy.waitForTinyMCELoaded().then(() => {
            cy.typeInTinyMCE('New content...');
        })

```
The new content is correctly seen, but it is not saved, as, I suppose, Directus is not aware of the modification.
Is there a way to add the field to the Directus list of the modified fields to get it properly saved on clicking the Save button?

**Comment by rijkvanzanten** [comment]
I don't know what either of those methods do, but it sounds like it's not triggering the mutation observer that's used to watch changes in the tinymce context

**Comment by Pachat** [comment]
Thank you for the insight.
The problem is that everything looks OK on the tinyMCE side:
The method `typeInTinyMCE` is
```
Cypress.Commands.add('typeInTinyMCE', (content) => {
    cy.window().its('tinyMCE').its('activeEditor').then(activeEditor => {
        activeEditor.setContent(content)
        activeEditor.save()
    })
})

```
and the tinyMCE [docs](https://www.tiny.cloud/docs-4x/api/tinymce/tinymce.editor/) states
>**setContent()** | Sets  the specified content to the editor instance, this will cleanup the  content before it gets set using the different cleanup rules options.

>**save()** | Saves  the contents from a editor out to the textarea or div element that got  converted into an editor instance. This method will move the HTML  contents from the editor into that textarea or div by getContent **so all  events etc that method has will get dispatched as well**.

This makes me feel that the event of modification is fired from the tinyMCE side.  Is it listened to by Directus?

**Reply by rijkvanzanten** [reply]
It emits on the set-content event afaict yes: 
https://github.com/directus/directus/blob/0bffb2b0f5a56449434cd34bc6e96109b0656bea/app/src/interfaces/input-rich-text-html/input-rich-text-html.vue#L458

**Reply by Pachat** [reply]
Thank you for the hint.
For some reason,  looks like `function contentUpdated()` is running only when typing by hand in the editor,
not when the editor's content is changed programmatically.

**Reply by licitdev** [reply]
@Pachat Currently, the `set-content` event and the mutation observer triggers `contentUpdated()`.

Perhaps we'll need to trigger `contentUpdated()` for other event(s). 🤔 
Ref: https://www.tiny.cloud/docs/integrations/vue/#eventbinding

https://github.com/directus/directus/blob/3e5676672018e57cd7d567c3121572bdb3dd8269/app/src/interfaces/input-rich-text-html/input-rich-text-html.vue#L3-L13

**Reply by Pachat** [reply]
> @licitdev Perhaps we'll need to trigger `contentUpdated()` for other event(s).

saveContent?

**Reply by rijkvanzanten** [reply]
@licitdev I'd rather not start hacking in more events that might cause other issues in the app to play ball with an external automated suite that we're not really relying on. In this case, I'd recommend to update the test suite to send some keyboard commands into the frame rather than relying on the setContent command (which we're already listening to regardless, indicating that this could also be a tinymce-vue event listening problem, or the activeEditor being the wrong instance)

**Reply by Pachat** [reply]
I understand. And had already tried that but it did not work.
The problem with Cypress is that they currently do not fully support iframes which makes it difficult to enter inside an iframe.
I'll wait until they get better on this.
